azkaban-common/src/main/java/azkaban/executor/container/ContainerCleanupManager.java

```diff
@@ -325,9 +343,45 @@ private void cleanUpYarnApplicationsInCluster(Set<Integer> toBeCleanedContainers
       return;
     }
 
-    // todo: kill the apps asynchronously
+    // Use a fix thread pool to concurrently kill yarn apps
+    Map<String, Boolean> appsSuccessfulKilled = new ConcurrentHashMap<>();
+    aliveApplications.forEach(
+        app -> appsSuccessfulKilled.put(app.getApplicationId().toString(), false));
+
+    ExecutorService yarnKillThreadPool = Executors.newFixedThreadPool(this.yarnAppKillParallelism);
```

Comment:
Two questions:
1. Will there be two cleanUpYarnApplicationsInCluster instances running in parallel? If yes, we will spin up too many threads for yarn kill purpose: this.yarnAppKillParallelism * number_of_cleanUpYarnApplicationsInCluster_parallelism.
2. Can you move yarnKillThreadPool to a global var instead of a local var? I feel like creating a thread pool is heavy.

Reply:
1. No it won't, looking at line 322, `cleanUpYarnApplicationsInCluster` will be called sequentially for multiple clusters.
2. good point, will do

Reply:
Ah, sorry. Let me elaborate more on the first question.
Will two cleanUpDanglingYarnApplications instances be executed in parallel (if yes, cleanUpYarnApplicationsInCluster will be executed in parallel as well)? cleanUpDanglingYarnApplications is called by scheduleAtFixedRate. Suppose the first run of cleanUpDanglingYarnApplications lasts for 1 day, and then in the middle of that 1d, will scheduleAtFixedRate call cleanUpDanglingYarnApplications again or simply wait for the 1st run to be completed?

Reply:
From the method signature itself:
`If any execution of this task takes longer than its period, then subsequent executions may start late, but will not concurrently execute.`